### ISSUE TITLE
Fitness Test or Timed Walk does not work when app is in the background

### ISSUE THREAD

**Issue by duncangroenewald** [ResearchKit/ResearchKit#1176]
Is there any plan to address the issue with running a timer in the background for some of these tasks.  Does anyone else think they are all but useless without this capability as it seems impractical to expect the user to keep the app in the foreground - and there are no such instructions in the ResearchKit tasks currently.

If nothing else wouldn't it be useful to add UserNotifications to alert the user at the appropriate time ?

Anyone aware of any attempts to address this issue ?

**Comment by MittaKarthik** [comment]
Could you please help in providing the solution for this issue @srinathtm-apple.

The app is terminating if it goes to the background or if the phone locked while performing the 6 Min walk test.

**Comment by davwillev** [comment]
I think active tasks that perform some (time-limited, perhaps up to 1 hour) background sensor recording in general is an unavoidable necessity at some point, even if users are required to initiate the recording.

**Comment by MittaKarthik** [comment]
Yes @davwillev. As you know apple reduced the background execution time from 10 mins to 6 mins to now in iOS 13 it's 30 seconds (I know it's not soo clear in their document), OS is trying to look for proper UIBackgroundtask handling but it's not handled properly for 6 Min walk test, the watchdog is killing the app. This is what I understood from the code why it's crashing.

I am surprised why Apple didn't think about the app can go to the background when the user perform walk test, becuase it's not correct to ask users to perform in the foreground. Not sure is there any hidden reason behind this.

BTW, did you also faced this issue in your application by any chance?

**Comment by duncangroenewald** [comment]
You can try working around this by using core location as this will keep the app alive when it's backgrounded.  I am using location to estimate the distance - as I recall I have to substantially change the ResearchKit code.

**Comment by MittaKarthik** [comment]
@duncangroenewald  But there is a chance of rejection if we use always location allow in background for these kinds of applications right?

Could you please help me the code change which you are referring to.

**Comment by duncangroenewald** [comment]
You won't get rejected as long as you explain the reason for using location - and given you are measuring distance and speed during the timed walk it's unlikely they will reject your app.  It will be a lot of work for me to find the code changes as I have not worked on that app for some years now.  But it went something like this:
- when the timed walk starts start location tracking and use the GPS coordinates to determine the distance travelled and the speed (optional of course).
- Also start a timer or set a timed notification (something like that) so the user gets an alert when the time runs out - maybe a countdown alert of some kind - say every minute - you should check if the app is in the background to determine the type of notification to post to the user.
- Also check the time when you get the Location callbacks and if necessary post a notification to the user

- I can't recall the exact details but one of my apps locks the screen ON so you can do that as well - but I wouldn't recommend that approach.  Mine is for a specific training scenario - and I had to send Apple a video of the application in use for that scenario before they would approve it - where the phone has an external power source and provides a continuous heads up display - a bit like a maps does when used while driving a car.
